src/Hadolint/Rules.hs

```diff
@@ -1284,3 +1292,178 @@ useraddFlagL = instructionRule code severity message check
     hasLFlag = Shell.hasAnyFlag ["l", "no-log-init"]
     hasUFlag = Shell.hasAnyFlag ["u", "uid"]
     hasLongUID cmd = any ((> 5) . Text.length) (Shell.getFlagArg "u" cmd)
+
+noInvalidLabelKey :: Rule
+noInvalidLabelKey = instructionRule code severity message check
+  where
+    code = "DL3047"
+    severity = DLStyleC
+    message = "Invalid label key."
+    check (Label pairs) = hasNoInvalidKey pairs
+    check _ = True
+
+    hasNoInvalidKey prs = null $ [(l, v) | (l, v) <- prs,
+                                          Text.head l `notElem` ['a'..'z'] ||
+                                          Text.last l `notElem` ['a'..'z'] ++ ['0'..'9'] ||
+                                          containsIllegalChar l ||
+                                          hasReservedNamespace l ||
+                                          hasConsecutiveSeparators l]
+    containsIllegalChar :: Text.Text -> Bool
+    containsIllegalChar l = any (`notElem` validChars) (Text.unpack l)
```

Comment:
This could be ```Text.any (\c -> c `notElem` validChars)```

Reply:
Even easier than that, `Text.any (\`notElem\` validChars)` works just fine.